src/backend/distributed/worker/worker_shard_visibility.c

```diff
@@ -112,10 +113,38 @@ ErrorIfRelationIsAKnownShard(Oid relationId)
 	}
 
 	const char *relationName = get_rel_name(relationId);
+
 	ereport(ERROR, (errmsg("relation \"%s\" is a shard relation ", relationName)));
 }
 
 
+/*
+ * ErrorIfChangingKnownShard errors out if the relation with relationId is
+ * a known shard and manual changes on known shards are disabled. This is
+ * valid for only non-citus (external) connections.
+ */
+void
+ErrorIfChangingKnownShard(Oid relationId)
+{
+	if (IsCitusInitiatedRemoteBackend() || EnableManualChangesToShards)
```

Comment:
Hmz, does IsCitusInitiatedRemoteBackend return false for local execution? I'm surprised we're not hitting this.

Reply:
Yes, for local execution it returns false. So when `EnableManualChangesToShards` is also false (by default) we never go into that if statement

Reply:
It seems that the reason Citus does not hit the error when dropping/truncating shards via local execution is that RelationIsAKnownShard somewhat accidentally (e.g. because of the search_path in the drop trigger, or because of the distributed table being dropped) returns false.

It seems quite fragile. We should probably have another check that determines whether we are in a "local execution".

Reply:
Even if we know whether it's a local execution, we should also know that the command belongs to the user or Citus